core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -481,35 +546,41 @@ class TiBatchWrite(@transient val df: DataFrame,
     }
   }
 
-  // currently deduplicate can only perform on pk is handle table.
   @throws(classOf[TiBatchWriteException])
-  private def deduplicateIfNecessary(rdd: RDD[TiRow], necessary: Boolean): RDD[TiRow] =
-    if (necessary) {
-      val shuffledRDD = rdd.map(row => (tiKVRow2Key(row), row))
-      val rddGroupByKey = shuffledRDD.groupByKey
-      if (!options.deduplicate) {
-        val duplicateCountRDD = rddGroupByKey.flatMap {
-          case (_, iterable) =>
-            if (iterable.size > 1) {
-              Some(iterable.size)
-            } else {
-              None
-            }
-        }
-
-        if (!duplicateCountRDD.isEmpty()) {
-          throw new TiBatchWriteException("data conflicts! set the parameter deduplicate.")
+  private def deduplicate(rdd: RDD[WrappedRow]) = {
+    //1 handle key
+    var mutableRdd = rdd
+    if (handleCol != null) {
+      mutableRdd = mutableRdd
+        .map { wrappedRow =>
+          val rowKey = buildRowKey(wrappedRow.row, wrappedRow.handle)
+          (rowKey, wrappedRow)
         }
+        .groupByKey()
+        .map(_._2.head)
+    }
+    uniqueIndices.foreach { index =>
+      {
+        mutableRdd = mutableRdd
+          .map { wrappedRow =>
+            val indexKey = buildUniqueIndexKey(wrappedRow.row, index)
+            (indexKey, wrappedRow)
+          }
+          .groupByKey()
+          .map(_._2.head)
       }
-
-      rddGroupByKey.map {
-        case (_, iterable) =>
-          // remove duplicate rows if key equals
-          iterable.head
-      }
+    }
+    if (options.deduplicate) {
```

Comment:
suggest to use `options.deduplicate || options.replace` since when in replace mode, user usually do not expect data conflicts error

Reply:
In that case, we can actually remove `deduplicate` option. @marsishandsome What do you think? As far as I can recall, you think we need `deduplicate` option and let users choose to set deduplicate or not.